### ISSUE TITLE
Decorator returning a callback protocol does not behave like returning a Callable

### ISSUE THREAD

**Issue by moosbruggerj** [microsoft/pyright#3969]
**Describe the bug**
If a decorator returns a Callback Protocol instead of a Callable, a decorated method raises an error when calling.
It reports one missing parameter, although it works at runtime.

The revealed type for `Test().test` at runtime is `method` as expected.

**To Reproduce**
See code below.

If the decorator is defined with the commented out declaration, the code snipped type checks and revealed type of `Test().test` is `() -> None` as expected.

**Expected behavior**
The code should type check correctly with the protocol, and reveal the same type for function `Test().test` as the version with `Callback`.
According to [PEP 544](https://peps.python.org/pep-0544/#callback-protocols) Callback Protocols should be interchangeable with Callback.

**Screenshots or Code**

```python
from typing import Protocol, Callable, TypeVar, Generic
from typing_extensions import reveal_type
T = TypeVar('T')

class Func(Protocol, Generic[T]):
    def __call__(self, _: T, /) -> None:
        ...

# def decorator(f: Callable[[T], None]) -> Callable[[T], None]:
def decorator(f: Callable[[T], None]) -> Func[T]:
    def _wrapper(self: T) -> None:
        f(self)

    return _wrapper

class Test:
    @decorator
    def test(self) -> None:
        pass

Test().test() # Expected 1 more positional argument (Pyright reportGeneralTypeIssues)
reveal_type(Test().test) # reveals Func[Test]

```

**VS Code extension or command-line**
Nvim: coc-pyright 1.1.271 with Pyright 1.1.271

**Comment by erictraut** [comment]
The current behavior is correct. For comparison, mypy gives the same error in this case.

The problem is that you're decorating an instance method — one that normally requires binding to an object before it can be called. There is no way in the current Python type system to describe the difference between a callable that needs to be bound to an object (an instance method), a callable that needs to be bound to a class (a class method), or a callable that doesn't require binding (a static method, normal function, or an already-bound method). The transform applied by the `decorator` function results in a `Func` object, and this loses the notion that the captured call is an instance method that requires binding.

How does this work with a `Callable`? That's handled through special-case logic in pyright that retains the internal flags that indicate that the resulting callable is an instance method. This behavior is not specified in PEP 484 or any of the Python typing docs, but all Python type checkers have implemented it out of convenience specifically to handle decorators like this. Such special-case logic can't be applied generally to a callback protocol because the signature of the callable within a callback protocol is spelled out explicitly in the provided `__call__` method.

Is there a reason why the `Callable` doesn't work for you in this case? I'm guessing that there was some reason that led you to use a callback protocol rather than sticking with the simpler `Callable` approach.

**Comment by moosbruggerj** [comment]
Thank you for the detailed explanation, I was not aware of the special casing for `Callable`.

In my use case, I have a decorator, that adds some keyword only arguments. I know, this is not exactly type save, as outlined in the `ParamSpec` PEP. However, using position only arguments would complicate other methods in the class hierarchy and make their calling conventions slightly different, which is annoying.
But I therefore cannot use a `ParamSpec`.
I also do not want to use an unbound `Callable[..., T]` signature, as those functions are generally factory methods returning `Self`.
An unbound callable, makes the `Self` type unbound, breaking type hinting for subclasses, if I don't explicitly overwrite and annotate the return types of the factory functions.

That's why I tried using the protocols.
Another possiblity I am currently trying is using `*tuple[Any, ...]` from PEP 646 for any real parameters, so I can at least correctly annotate the `self/cls` argument and can still bind the `Self` return value.
But there I run into a similar issue, with missing parameters:
```python
def decorator(f: Callable[[T], None]) -> Callable[[T, Unpack[tuple[int, ...]]], None]:
    def _wrapper(self: T, *args: int) -> None:
        print(args)
        f(self)

    return _wrapper


class Test:
    @decorator
    def test(self) -> None:
        pass


Test().test() # Expected 1 more positional argument (Pyright reportGeneralTypeIssues)
reveal_type(Test().test) # Type of "Test().test" is "(*tuple[int, ...]) -> None"
```
